query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java

```diff
@@ -171,61 +168,42 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
-      }
-
-      // If the column names of query filter and view filter match but the filters
-      // don't match, we can exit here.
-      String viewFilterColumnName = viewFilterMap.keySet().stream().findFirst().get();
-      if (filter.getLhs().getValueCase() == ValueCase.COLUMNIDENTIFIER &&
-          StringUtils.equals(viewFilterColumnName,
-              filter.getLhs().getColumnIdentifier().getColumnName()))
-      {
-        return doesSingleViewFilterMatchLeafQueryFilter(viewFilterMap.get(viewFilterColumnName),
-            filter);
-      }
-
-      return false;
+      return doesSingleViewFilterMatchLeafQueryFilter(viewFilterMap, filter);
     } else {
       Map<String, Filter> childLeafFilterMap = filter.getChildFilterList().stream()
-          .filter(f -> f.getChildFilterCount() == 0 &&
-              f.getLhs().getValueCase() == Expression.ValueCase.COLUMNIDENTIFIER)
-          .filter(
-              f -> viewFilterMap.containsKey(f.getLhs().getColumnIdentifier().getColumnName()))
+          .filter(f -> f.getChildFilterCount() == 0)
           .collect(Collectors
               .toMap(f -> f.getLhs().getColumnIdentifier().getColumnName(), f -> f));
+
+      // 2. An internal filter node which doesn't have any leaves but other internal nodes:
+      // Parent is a match if all the OR'ed children match or at least one of AND'ed children match.
       if (childLeafFilterMap.isEmpty()) {
         long matchCount = filter.getChildFilterList().stream()
             .map(f -> doViewFiltersMatch(f, viewFilterMap))
             .filter(b -> b).count();
-        if (filter.getOperator() == Operator.AND && matchCount > 0) {
-          return true;
-        }
 
-        // When children are being OR'ed, all the children must match the view filters.
-        return filter.getOperator() == Operator.OR && matchCount == filter.getChildFilterCount();
-      }
-
-      if (filter.getOperator() != Operator.AND)
-        return false;
+        return (filter.getOperator() == Operator.AND && matchCount > 0) ||
+            (filter.getOperator() == Operator.OR && matchCount == filter.getChildFilterCount());
+      } else {
+        // 3. An internal filter with children which are leaves: All the view filters
+        // should be matched and AND'ed, otherwise it's not a match.
+        long matchCount = filter.getChildFilterList().stream()
+            .map(f -> doViewFiltersMatch(f, viewFilterMap))
+            .filter(b -> b).count();
 
-      for (Map.Entry<String, ViewColumnFilter> entry : viewFilterMap.entrySet()) {
-        // Make sure all view filters match.
-        Filter queryFilter = childLeafFilterMap.get(entry.getKey());
-        if (queryFilter == null || !doesSingleViewFilterMatchLeafQueryFilter(entry.getValue(), queryFilter)) {
-          // Few columns matched but others didn't match.
-          return false;
-        }
+        return filter.getOperator() == Operator.AND && matchCount == viewFilterMap.size();
```

Comment:
This is checking for a "full match" but misses a number of scenarios. All we know about these children is that at least one is a leaf, but not necessarily all.
In the AND case, we need each view filter to be matched, but that could happen in one or more nodes. It's also not necessarily indicated by the count as we could have duplicates:

Duplicate case:
View Filter: X = 'A' AND Y = 'B'
Query Filter: X = 'A' AND X= 'A'

Single match case (this specific example wouldn't happen given our optimize logic, but are we intending on relying on that here? that seems like a far distance for an assumption to travel):
View Filter: X = 'A' AND Y = 'B'
Query Filter: (X = 'A' AND Y= 'B') AND Z = 'C'

In the OR case, we need each child to be a full match, but right now we'd never match since we only look at ANDs

Each full match case:
View Filter: X = 'A'
Query Filter: X = 'A' OR X = 'A' // many equivalencies to this

Reply:
Yes, I'm relying on the filter optimization for all these 3 cases and other similar duplicate cases or flattening predicates. I feel that keeps the tree navigation logic much simpler instead of handling so many cases.
Reg keeping the filter optimization far from this code: I thought about it and the reason I'm doing it early on is 
1. so that we don't have to repeat it in each handler
2. other parts of the code like conversion to SQL can take advantage of it too.

The optimization logic is in a separate class and can be invoked from here too if needed but I think this is fine.

Reply:
See the later part of my most recent comment - https://github.com/hypertrace/query-service/pull/17#discussion_r489412091  - I believe this proposes a way of doing this in a much more straightforward way that doesn't require (but works with) optimization and all scenarios.